pkg/services/ngalert/state/historian/annotation.go

```diff
@@ -31,48 +31,60 @@ func NewAnnotationHistorian(annotations annotations.Repository, dashboards dashb
 	}
 }
 
-func (h *AnnotationStateHistorian) RecordState(ctx context.Context, rule *ngmodels.AlertRule, labels data.Labels, evaluatedAt time.Time, currentData, previousData state.InstanceStateAndReason) {
-	h.log.Debug("alert state changed creating annotation", "alertRuleUID", rule.UID, "newState", currentData.String(), "oldState", previousData.String())
-
-	labels = removePrivateLabels(labels)
-	annotationText := fmt.Sprintf("%s {%s} - %s", rule.Title, labels.String(), currentData.String())
+func (h *AnnotationStateHistorian) RecordStates(ctx context.Context, states []state.ContextualState) {
+	go h.recordStatesSync(ctx, states)
```

Comment:
The state is a pointer and technically it can change. Therefore we process synchronously everywhere in rule evaluation routine relying to the fact that nothing else processes states for that rule concurrently. In this case we do this asynchronously which can potentially be unsafe.

I would recommend a slightly different approach:
```go
func (h *AnnotationStateHistorian) RecordStates(ctx context.Context, rule AlertRule, states []state.ContextualState) {
     var items := convertStatesToItems(rule, states) // converts without populating DashboardId\PanelId
     go recordItems(rule.DashboardUID, rule.PanelId, items)
}
```

this case we convert states to annotations synchronously and then populate with data async

Reply:
Good call - this got a little blurred when I moved ContextualState to wrap a pointer rather than copy. But, that approach introduced similar issues in different places :D

I guess there really is no way out here other than copying data out ourselves.